pkg/ccl/backupccl/restore_job.go

```diff
@@ -651,6 +650,28 @@ func spansForAllRestoreTableIndexes(
 	return spans
 }
 
+// findLatestIntroFromManifests finds the endtime of the latest incremental
+// backup that introduced each backed up index, as of restore time.
+func findLatestIntroFromManifests(
+	manifests []backuppb.BackupManifest, codec keys.SQLCodec, asOf hlc.Timestamp,
+) (map[tableAndIndex]hlc.Timestamp, error) {
+	latestIntro := make(map[tableAndIndex]hlc.Timestamp)
+	for _, b := range manifests {
+		if !asOf.IsEmpty() && asOf.Less(b.StartTime) {
+			break
+		}
+		for _, sp := range b.IntroducedSpans {
+			_, tablePrefix, indexPrefix, err := codec.DecodeIndexPrefix(sp.Key)
+			if err != nil {
+				return nil, err
+			}
+			introKey := tableAndIndex{descpb.ID(tablePrefix), descpb.IndexID(indexPrefix)}
```

Comment:
I think a lot of other backup and restore code would break if {tableID,IndexID} didn't 1:1 map to a unique backup or restore span. `roachpb.MergeSpans()` should never merge two spans from different indexes, right? We seem to rely on this invariant via the `tableAndIndex` struct throughout the backup and restore code.

Reply:
>  should never merge two spans from different indexes

I think it merges contiguous spans even if they're from different indexes so [/table/12/1 - /table/12/2) and [/table/12/2 - /table/12/3) will merge to [/table/12/1-/table/12/3), and [/table/12/1-/table/12/2) + [/table/13/1-/table/13/2) merges to [/table/12/1 - /table/13/2)? That aside I do agree that we do rely on {tableID, indexID} elsewhere but is it correct to create a mapping based on `IntroducedSpans` created during backup and then elide **all** spans with the same {tableID, indexID} during restore. You might be right I just want to convince myself, so assume when taking an incremental backup we see table 12 comes online so `IntroducedSpans` will have [/table/12/1-/table/13/1) so when restoring we will run this check for spans where the start key is in table 12 and index 1 but what about table 12 and index 2?

Reply:
> I think it merges contiguous spans even if they're from different indexes so [/table/12/1 - /table/12/2) and [/table/12/2 - /table/12/3) will merge to [/table/12/1-/table/12/3), and [/table/12/1-/table/12/2) + [/table/13/1-/table/13/2) merges to [/table/12/1 - /table/13/2)

I don't think your examples are correct because the span for table 12, index 1, is [/table/12/1/0 - /table/12/1/{prefixEnd}), not [/table/12/1 - /table/12/2) -- i.e. the end key for a given index span will have the same prefix as the start key. Notice how index spans are created using a `PrefixEnd()` call within `forEachPublicIndexTableSpan` Given this, I don't _think_ spans from different indices overlap, and consequently merge.

Reply:
Hmm you're right about the prefixEnd but I did the following:

```
CREATE TABLE foo (id INT PRIMARY KEY);
CREATE INDEX bar ON foo (id);
INSERT INTO foo VALUES (1);

root@localhost:26257/defaultdb> select crdb_internal.pretty_key(key, 0) from crdb_internal.scan(crdb_internal.table_span(104));
  crdb_internal.pretty_key
----------------------------
  /104/1/1/0
  /104/2/1/0
(2 rows)

root@localhost:26257/defaultdb> backup table foo into 'nodelocal://1/foo';
        job_id       |  status   | fraction_completed | rows | index_entries | bytes
---------------------+-----------+--------------------+------+---------------+--------
  795137589190295553 | succeeded |                  1 |    1 |             1 |    18
(1 row)
```

I added a print statement before and after `roachpb.MergeSpans()` in `spansForAllTableIndexes` and it shows me:
```
I220909 12:43:44.609424 2252 ccl/backupccl/backup_planning.go:174 ⋮ [n1,job=795138537598124033] 219  these are the unmerged spans [‹/Table/104/{1-2}› ‹/Table/104/{2-3}›]
I220909 12:43:44.609567 2252 ccl/backupccl/backup_planning.go:181 ⋮ [n1,job=795138537598124033] 220  these are the merged spans [‹/Table/104/{1-3}›]
```

Reply:
wow, great catch! It's quite subtle that the PrefixEnd (i.e. the endKey) of an index span can be the start key of a different index which causes index spans to merge; whereas that's not the case for table spans. I'll add a bit more documentation to `forEachPublicIndexTableSpan`, maintain a tableID map instead, and add some indices in my tests.